In [1]:
import pandas as pd

In [2]:
data = {
    "Age" : [20,30,34],
    "profession" : ["student","student","teacher"]
}

In [3]:
df = pd.DataFrame(data, index = ["Ram", "Shyam", "Hari"])

In [4]:
df

,Age,profession
Ram,20,student
Shyam,30,student
Hari,34,teacher


In [5]:
df.loc["Ram"] # row

,Ram
Age,20
profession,student


In [6]:
df["Age"]# column

,Age
Ram,20
Shyam,30
Hari,34


In [7]:
# slicing multiple rows
df.loc["Ram":"Shyam"]

,Age,profession
Ram,20,student
Shyam,30,student


In [8]:
df.iloc[1:3]

,Age,profession
Shyam,30,student
Hari,34,teacher


In [9]:
#Row and Colum Selection combined
# .loc: Specify row and column labels directly
df.loc['Ram','Age']

np.int64(20)

In [10]:
# Using .iloc: Use integer indices for both dimensions.
df.iloc[0,1]

'student'

In [11]:
df.iloc[2,1]

'teacher'

In [12]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [13]:
titanic_df = pd.read_csv("/content/drive/MyDrive/Data Science and Machine Learning/titanic-data.csv")

In [14]:
titanic_df.shape # gives dimension

(891, 12)

In [15]:
titanic_df.columns # shows all the column

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [16]:
titanic_df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [17]:
titanic_df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [18]:
titanic_df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [19]:
titanic_df.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


In [20]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [21]:
titanic_df[["Age", "Fare"]].mean() # finding the mean

,0
Age,29.699118
Fare,32.204208


In [22]:
titanic_df[["Age", "Fare"]].describe()

,Age,Fare
count,714.000000,891.000000
mean,29.699118,32.204208
std,14.526497,49.693429
min,0.420000,0.000000
25%,20.125000,7.910400
50%,28.000000,14.454200
75%,38.000000,31.000000
max,80.000000,512.329200


**Description:**
    
* Average age of passengers is approximately 29.7 years, indicating that most passengers were relatively young.
* Age distribution is fairly symmetric, as the mean (29.7) and median (28.0) are close.
* The minimum age is 0.42 years (about 5 months), and the oldest passenger is 80 years old.
* There are missing values in the Age column (714 entries vs. 891 for Fare), suggesting some age data is not available.

* For Fare, the mean is 32.20 and the median is 14.45 — a large gap indicating a right-skewed (positively skewed) distribution.
* Fare values range from 0 (free tickets) to over 512, showing a wide spread and presence of outliers or high-paying passengers.
* The high standard deviation (49.69) also supports the presence of large variability in fare prices.


In [23]:
titanic_df["Survived"].value_counts()

,count
Survived,
0,549
1,342


In [24]:
titanic_df.value_counts("Sex")

,count
Sex,
male,577
female,314


In [25]:
titanic_df.value_counts(["Sex", "Survived"], ascending=True)

,,count
Sex,Survived,
female,0,81
male,1,109
female,1,233
male,0,468


In [26]:
titanic_df.value_counts(["Sex", "Survived"], ascending=True, normalize=True)

,,proportion
Sex,Survived,
female,0,0.090909
male,1,0.122334
female,1,0.261504
male,0,0.525253


In [27]:
titanic_df[titanic_df["Sex"]=="female"].value_counts("Survived",ascending=True, normalize=True)

,proportion
Survived,
0,0.257962
1,0.742038


Building a survival profile

In [35]:
# analyzing survival by zender
gender_survival = titanic_df.groupby("Sex")["Survived"].mean() * 100

print("Survival by Gender:")
print(gender_survival)

Survival by Gender:
Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64


In [36]:
# survival by passenger class
class_survival = titanic_df.groupby("Pclass")["Survived"].mean() * 100

print("\nSurvival by Passenger Class:")
print(class_survival)


Survival by Passenger Class:
Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64


In [42]:
# creating the age grpup
titanic_df["AgeGroup"] = pd.cut(
    titanic_df["Age"],
    bins=[0, 12, 18, 35, 60, 100],
    labels=["Child", "Teenager", "Young Adult", "Adult", "Senior"]
)

In [45]:
# sruvival by age group
age_survival = titanic_df.groupby("AgeGroup", observed=True)["Survived"].mean() * 100

print("\nSurvival by Age Group:")
print(age_survival)


Survival by Age Group:
AgeGroup
Child          57.971014
Teenager       42.857143
Young Adult    38.268156
Adult          40.000000
Senior         22.727273
Name: Survived, dtype: float64


In [61]:
# creating the fare group with bins
titanic_df["FareGroup"] = pd.cut(
    titanic_df["Fare"],
    bins=[0, 10, 20, 30, 100],
    labels=["Low", "Medium", "High", "Very High"]
)

print(titanic_df[["Fare", "FareGroup"]].head())
fare_sruvival = titanic_df.groupby("FareGroup", observed = True)["Survived"].mean() * 100

print("\nSurvival by Fare Group:")
print(fare_sruvival)

      Fare  FareGroup
0   7.2500        Low
1  71.2833  Very High
2   7.9250        Low
3  53.1000  Very High
4   8.0500        Low

Survival by Fare Group:
FareGroup
Low          20.560748
Medium       42.458101
High         44.366197
Very High    53.591160
Name: Survived, dtype: float64


In [56]:
# finding the strongest commbined profile
profile = (
    titanic_df.groupby(["Sex", "Pclass", "AgeGroup", "FareGroup"], observed = True)
    .agg(
        sruvival_rate = ("Survived", "mean"),
        Passenger_count = ("Survived", "count")
    )
    .sort_values(by = "sruvival_rate", ascending = False)
    )

profile["survival_rate"] = profile["sruvival_rate"] * 100
profile.head(10)

sruvival_rate  Passenger_count  \
Sex    Pclass AgeGroup    FareGroup                                   
female 1      Teenager    Very High            1.0                5   
              Young Adult High                 1.0                2   
                          Very High            1.0               22   
              Adult       Very High            1.0               18   
              Senior      Very High            1.0                2   
       2      Child       High                 1.0                5   
       3      Senior      Low                  1.0                1   
       2      Child       Very High            1.0                3   
              Teenager    Medium               1.0                4   
                          High                 1.0                1   

                                     survival_rate  
Sex    Pclass AgeGroup    FareGroup                 
female 1      Teenager    Very High          100.0  
              Young Adult High               100.0  
                          Very High          100.0  
              Adult       Very High          100.0  
              Senior      Very High          100.0  
       2      Child       High               100.0  
       3      Senior      Low                100.0  
       2      Child       Very High          100.0  
              Teenager    Medium             100.0  
                          High               100.0

In [57]:
reliable_profiles = profile[
    profile["Passenger_count"] >= 10
]

print(reliable_profiles.head(10))

                                     sruvival_rate  Passenger_count  \
Sex    Pclass AgeGroup    FareGroup                                   
female 1      Young Adult Very High       1.000000               22   
              Adult       Very High       1.000000               18   
       2      Young Adult Medium          0.944444               18   
                          High            0.894737               19   
              Adult       Medium          0.833333               12   
       3      Child       Medium          0.818182               11   
              Teenager    Low             0.750000               12   
              Young Adult Medium          0.571429               14   
male   1      Young Adult Very High       0.529412               17   
female 3      Young Adult Low             0.480000               25   

                                     survival_rate  
Sex    Pclass AgeGroup    FareGroup                 
female 1      Young Adult Very High     1

Embarkation prot analysis

In [62]:
# counting passenger from each prot
port_counts = titanic_df["Embarked"].value_counts()

print("Passenger Count by Port of Embarkation:")
print(port_counts)

Passenger Count by Port of Embarkation:
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64


In [63]:
# calculating the survival rate for each prot
port_sruvival = titanic_df.groupby("Embarked")["Survived"].mean() * 100

print("\nSurvival Rate by Port of Embarkation:")
print(port_sruvival)



Survival Rate by Port of Embarkation:
Embarked
C    55.357143
Q    38.961039
S    33.695652
Name: Survived, dtype: float64


In [65]:
# Finding the port with highest survival rate
highest_survival_port = port_sruvival.idxmax()
highest_survival_rate = port_sruvival.max()

print("\nPort with Highest Survival Rate:", highest_survival_port)
print("Survival Rate:", highest_survival_rate)


Port with Highest Survival Rate: C
Survival Rate: 55.35714285714286
